src/main/java/com/microsoft/graph/authentication/TokenCredentialAuthProvider.java

```diff
@@ -0,0 +1,141 @@
+    /** maximum delay to wait for token obtention */
+    private final Duration tokenBlockTimeout;
+    /** Default scope to use when no scopes are provided */
+    private static final String DEFAULT_GRAPH_SCOPE = "https://graph.microsoft.com/.default";
+    /** Default timeout for token obtention, set to 10 minutes in case of interactive flows to give time to the end user */
+    private static final long DEFAULT_TOKEN_TIMEOUT = 10;
+    /**
+     * Creates an Authentication provider using a passed in TokenCredential
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     */
+    public TokenCredentialAuthProvider(@Nonnull final TokenCredential tokenCredential) {
+        this(Arrays.asList(new String[] {DEFAULT_GRAPH_SCOPE}), tokenCredential);
+    }
+
+    /**
+     * Creates an Authentication provider using a TokenCredential and list of scopes with default timeout (10 minutes)
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @param scopes Specified desired scopes of the Auth Provider
+     */
+    public TokenCredentialAuthProvider(@Nonnull final List<String> scopes, @Nonnull final TokenCredential tokenCredential) {
+        this(scopes, Duration.ofMinutes(DEFAULT_TOKEN_TIMEOUT), tokenCredential);
+    }
+
+    /**
+     * Creates an Authentication provider using a TokenCredential and list of scopes
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @param scopes Specified desired scopes of the Auth Provider
+     * @param tokenObtentionTimeout Maximum time to wait for token obtention. Default 10 minutes. Use lower value on application with stable connectivity and no user interactions.
+     */
+    public TokenCredentialAuthProvider(@Nonnull final List<String> scopes, @Nonnull final Duration tokenObtentionTimeout, @Nonnull final TokenCredential tokenCredential) {
+        if(tokenCredential == null) {
+            throw new IllegalArgumentException("tokenCredential parameter cannot be null.");
+        }
+        if(tokenObtentionTimeout == null) {
+            throw new IllegalArgumentException("tokenObtentionTimout parameter cannot be null");
+        }
+        if(scopes == null || scopes.isEmpty()) {
+            throw new IllegalArgumentException("scopes parameter cannot be null or empty");
+        }
+        this.context = new TokenRequestContext();
+        this.context.setScopes(scopes);
+        this.tokenCredential = tokenCredential;
+        this.tokenBlockTimeout = tokenObtentionTimeout;
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     */
+    @Override
+    public void authenticateRequest(@Nonnull final IHttpRequest request) throws AuthenticationException {
```

Comment:
Can the logic to populate the AUTHORIZATION_HEADER be in a common place and the getAccessToken can have specific implementations depending on the authProvider?

Reply:
agreed, see my other comment. Let me know what you think about this new version